File: Dynamic_Programming/2318. Number of Distinct Roll Sequences.cpp
/*
2318. Number of Distinct Roll Sequences
*/

// TC:O(6*6*6*n) SC:O(n*6*6)
class Solution {
    using LL = long long;
    LL M = 1e9+7;
    LL dp[10001][7][7];
public:
    int distinctSequences(int n) {
        if(n == 1) return 6;

        LL count = 0; //處理 n == 2狀況
        for(int a = 1; a <= 6; a++){
            for(int b = 1; b <= 6; b++){
                if(a != b && gcd(a,b) == 1){
                    dp[2][a][b] = 1;
                    count = (count + dp[2][a][b])%M;
                }
            }
        }
        if(n == 2) return count;

        for(int i = 3; i <= n; i++){
            for(int a = 1; a <= 6; a++){
                for(int b = 1; b <= 6; b++){
                    if(a == b || gcd(a,b) > 1) continue;
                    for(int c = 1; c <= 6; c++){
                        if(c != a){
                            dp[i][b][a] += dp[i-1][c][b];
                            dp[i][b][a] %= M;
                        }
                    }
                }
            }
        }

        LL res = 0;
        for(int a = 1; a <= 6; a++){
            for(int b = 1; b <= 6; b++){
                res = (res + dp[n][a][b])%M;
            }
        }

        return res;
    }
};

/*
擲骰子
GCD = 1
1. 相鄰兩元素不能相同
2. 橫跨的一個元素也不能相同
 X X X X X X X 
         c b a
1. b != a
2. gcd(b,a) == 1
3. c != a
dp[i][b][a] = dp[i-1][c][b] 缺什麼補什麼
*/

File: Dynamic_Programming/2318. Number of Distinct Roll Sequences.py
# TC:O(6*6*6*n) SC:O(n*6*6) 空間可以優化到 6 * n 但是偏難想
'''
MOD = 10 ** 9 + 7
@cache
def dfs(i:int, prev1:int, prev2:int) -> int:
    if i == 0:
        return 1
    ans = 0
    for j in range(1, 7):
        if j == prev2 or j == prev1 or gcd(j, prev1) != 1: continue
        ans = (ans + dfs(i - 1, j, prev1)) % MOD
    return ans
'''
class Solution:
    def distinctSequences(self, n: int) -> int:
        '''
        問說有多少種可能
        1. 每次擲骰子相鄰兩個GCD要是1 
        2. 相同元素需要間隔至少3 ex: 0 1 2 3
        n = 1e4
        看來需要上一個的狀態 以及前前一個前前前一個是啥
        1e4 * 6 * 6 * 6  * lga ~= 1e6 
        '''

        # 遞推 也可以放在全域 
        MOD = 10 ** 9 + 7
        if n == 1:
            return 6
        f = [[[0] * 6 for _ in range(6)] for _ in range(n + 1)]
        f[2] = [[int(j != i and gcd(j + 1, i + 1) == 1) for j in range(6)] for i in range(6)]
        for i in range(2, n):
            for j in range(6): # cur
                for last1 in range(6): # prev1
                    if last1 != j and gcd(last1 + 1, j + 1) == 1:
                        f[i + 1][j][last1] = (f[i + 1][j][last1] + sum(f[i][last1][last2] for last2 in range(6) if j != last2))% MOD
        return sum(sum(row) for row in f[n]) % MOD


        '''
        遞歸 寫在全域最快 
        return dfs(n, 7, 7)
        '''
